cabby/abstract.py

```diff
@@ -180,19 +184,24 @@ def _execute_request(self, request, uri=None, service_type=None):
         if self.jwt_url and self.username and self.password:
             if not self.jwt_token:
                 self.refresh_jwt_token(session=session)
-            session = dispatcher.set_jwt_token(session, self.jwt_token)
 
-        message = dispatcher.send_taxii_request(
-            session,
-            self._prepare_url(uri),
-            request,
-            taxii_binding=self.taxii_binding,
-            timeout=self.timeout)
-
-        return message
+        for attempt in (1, 2):
```

Comment:
An idea that came to my mind is using a function parameter instead of a loop. Up to you to decide whether it's an improvement though:

```
def prepare_generic_session(self, refresh_token_on_error = True):
  ...
  except UnsuccessfulStatusError as exc:
    if refresh_token_on_error and exc.status == libtaxii.ST_UNAUTHORIZED:
      self.refresh_jwt_token(session=session)
      return self.prepare_generic_session(False)
    raise
```

or

```
def prepare_generic_session(self, refresh_token_on_error = True):
  ...
  except UnsuccessfulStatusError as exc:
    if refresh_token_on_error and exc.status == libtaxii.ST_UNAUTHORIZED:
      self.jwt_token = None
      return self.prepare_generic_session(False)
    raise
```

Reply:
Using a partial function was considered, but I found the for loop just as simple and more readable.

Reply:
that would not solve the problem i think.

`.prepare_generic_session()` does _not_ do any network requests, so in order to act on a flag like ‘refresh_token_on_error’ there would be a need for a wrapper around every request to check each response, and maybe retry.

it turns out that this ‘._execute_request()’ is like that already: all requests flow through this code, so adding the check (and the retry) here seemed a solution with minimal impact on the rest of the code.

Reply:
@wbolster ah, my bad, I didn't that this is a different function (hidden diff lines), so what I was suggesting applies to `_execute_request` instead of `prepare_generic_session`

Reply:
ah yes, now i understand what you mean. while i agree this code may benefit from some refactoring (we considered extracting more helpers while looking at this issue), but i don't think this particular suggestion would help much in practice, because...

there are exactly 4 different places that invoke `._execute_request()` and each of them would want ‘auto retry if token expired’ behaviour (and would hence all pass the same flag), so unconditionally doing a retry would achieve the same.